### PR TITLE
Added consumes to ESProducers in TrackingTools/Producers

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Added consumes calls to ESProducers and switched them to use ESGetToken. 
In addition, stopped holding copies of the PSet and instead hold the necessary values.

This is meant as a technical change and should not cause any changes in behavior.

#### PR validation:

The code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
The RelVal failures are from a missing file in the repository which is not connected to this change.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by Dr15Jones** [comment]
As far as I can tell, the differences come from differences in error logger messages.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/27047	322037c
- technical
- jenkins tests pass

IMHO, the ES consumes syntax is quite baroque

**Comment by fabiocos** [comment]
@Dr15Jones following @slava77 comment, for my understanding: what is the advantage of having a lambda function embedded in the constructor initializer list?

File: TrackingTools/Producers/src/AnalyticalPropagatorESProducer.cc [hunk 1]
```diff
@@ -13,38 +18,46 @@
...
+  const double dphiCut_;
+  const PropagationDirection dir_;
+};
+
+AnalyticalPropagatorESProducer::AnalyticalPropagatorESProducer(const edm::ParameterSet& p)
+    : magToken_{setWhatProduced(this, p.getParameter<std::string>("ComponentName"))
+                    .consumesFrom<MagneticField, IdealMagneticFieldRecord>(
+                        edm::ESInputTag("", p.getParameter<std::string>("SimpleMagneticField")))},
+      dphiCut_{p.getParameter<double>("MaxDPhi")},
+      dir_{[](std::string const& pdir) {
```
Comment (VinInn) [inline TrackingTools/Producers/src/AnalyticalPropagatorESProducer.cc]: given the number of places where this happens
https://cmssdt.cern.ch/lxr/search?%21v=CMSSW_11_0_X_2019-06-04-2300&_filestring=.*.cc&_advanced=1&_string=%22oppositeToMomentum%22&_casesensitive=1
most probably is better to factorize it in a function somewhere, even in "PropagationDirection.h" itself...

**Comment by slava77** [comment]
> @Dr15Jones following @slava77 comment, for my understanding: what is the advantage of having a lambda function embedded in the constructor initializer list?

my comment was not about the lambdas used in initialization. These are not related to the ES.
I was commenting about chained  .setConsumes(x).setConsumes(y) and .consumesFrom alternatives.

**Comment by Dr15Jones** [comment]
> @Dr15Jones following @slava77 comment, for my understanding: what is the advantage of having a lambda function embedded in the constructor initializer list?

It avoided declaring a function that would only be used in one place in that file. This allowed the data member to be declared `const` which is very handy when possible.

**Comment by fabiocos** [comment]
+1

@Dr15Jones thanks, the syntax looks a bit convoluted but I see the purpose

**Comment by Dr15Jones** [comment]
> IMHO, the ES consumes syntax is quite baroque

Suggestions are welcome. The `consumesFrom` more closely matches what is done for the ED case. In the first round of adding _consume_ I found that `consumesFrom` was more of a hassle when dealing with mulitple data products being consumed by the same callback function. Hence `setConsumes` was born and we've found it to be much more convenient. With that being said, `setConsumes` can not be used in the initializer list.

**Comment by makortel** [comment]
Part of #26748

**Pull request merged**